### PR TITLE
Aggregate multiple columns behavior & parameter list support

### PR DESCRIPTION
**Pull request by mringler**
In a project, I have to propagate several aggregated columns across several tables using the AggregateColumn behavior. It is similar to having items, which are aggregated on their groups, which are aggregated on their owner. This works fine, except I noticed that the AggregateColumn behavior causes extensive update cascades when changing the aggregation source table. In my project, there are four aggregation columns on three tables, and a single insert, update or delete on the source table causes a whooping  32 subsequent updates. This has become a problem, particularly when deleting several items takes about a minute, when it should be a fraction of a second.
Reason for the update cascade is that AggregateColumn behavior triggers a save for each aggregation, which in turn triggers the full aggregation on the parent table, and so on, leading to exponential complexity.

I figured there is an easy solution for this: just have AggregateColumn do all the aggregations and only then trigger save, and we are back at linear complexity. In my project, that means 3 instead of 32 updates, and easy adjustment in case of delete. So that's fantastic, and easy to implement using Propel behavior.
However, I need Propel schema definition to supply a list of columns to the behavior, and this is currently not supported. Even if I wrote my own little behavior, the only way to get it working is with some funky business, like putting json into an xml parameter attribute. Bummer. If only there were parameter list support in the schema parser...

Long story short, I figured I can do it in Propel itself, for the sake of the blazing fast of everybody. Seems to work nicely, all the tests run through. Hope you can use it, so I don't have to monkey patch around the issue.

For reference, the behavior can be declared like this:
```xml
<behavior name="aggregate_multiple_columns" id="...">
    <parameter name="foreign_table" value="score"/>
    <parameter-list name="columns">
        <parameter-list-item>
            <parameter name="column_name" value="total_score" />
            <parameter name="expression" value="SUM(score)" />
        </parameter-list-item>
        <parameter-list-item>
            <parameter name="column_name" value="number_of_scores" />
            <parameter name="expression" value="COUNT(score)" />
        </parameter-list-item>
    </parameter-list>
</behavior>
``` 
If you have feedback, I am happy to hear it!
Thank you, best wishes

### CODE REVIEW

**Review by stereomon — CHANGES_REQUESTED** [review]
Only a couple small things that I would like to see updated. Please also make sure that your code is on top of the latest master.

File: src/Propel/Generator/Behavior/AggregateMultipleColumns/AggregateMultipleColumnsBehavior.php [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+
+    /**
+     * @return string
+     */
+    private function buildAggregationName()
+    {
+        $foreignTableName = $this->getForeignTable()->getPhpName();
+        $baseAggregationName = 'AggregatedColumnsFrom' . $foreignTableName;
+        $tableName = $this->getTable()->getPhpName();
+        if (!array_key_exists($tableName, self::$insertedAggregationNames)) {
```
[the remark this answers is not in the document]
Comment (mringler) [reply src/Propel/Generator/Behavior/AggregateMultipleColumns/AggregateMultipleColumnsBehavior.php]: ok, no problem

File: src/Propel/Generator/Behavior/AggregateMultipleColumns/AggregateMultipleColumnsBehavior.php [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+        }
+
+        $existingNames = &self::$insertedAggregationNames[$tableName];
+        if (!in_array($baseAggregationName, $existingNames)) {
+            $existingNames[] = $baseAggregationName;
+
+            return $baseAggregationName;
+        }
+
+        $ix = 1;
```
[the remark this answers is not in the document]
Comment (mringler) [reply src/Propel/Generator/Behavior/AggregateMultipleColumns/AggregateMultipleColumnsBehavior.php]: ok, no problem

File: src/Propel/Generator/Behavior/AggregateMultipleColumns/AggregateMultipleColumnsBehavior.php [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+            $this->aggregationName = $this->buildAggregationName();
+        }
+
+        return $this->aggregationName;
+    }
+
+    /**
+     * @return string
+     */
+    private function buildAggregationName()
```
[the remark this answers is not in the document]
Comment (mringler) [reply src/Propel/Generator/Behavior/AggregateMultipleColumns/AggregateMultipleColumnsBehavior.php]: Added type hints. I usually do, but thought you don't do it in Propel, as they are usually not present in the existing code. Knowing that you actually want return types, I'll add them again.

File: src/Propel/Generator/Behavior/AggregateMultipleColumns/AggregateMultipleColumnsBehavior.php [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+        return implode(', ', $selects);
+    }
+
+    /**
+     * @return string
+     */
+    protected function addObjectUpdate()
+    {
+        $table = $this->getTable();
+        $columnPhpNames = array_map(function (array $colDef) use ($table) {
```
[the remark this answers is not in the document]
Comment (mringler) [reply src/Propel/Generator/Behavior/AggregateMultipleColumns/AggregateMultipleColumnsBehavior.php]: ok, no problem

File: src/Propel/Generator/Builder/Util/SchemaReader.php [hunk 5]
```diff
@@ -483,4 +547,80 @@ protected function isAlreadyParsed($filePath)
     {
         return isset($this->schemasTagsStack[$filePath]);
     }
+
+    /**
+     * @param array $attributes attributes of parameter-list tag
+     *
+     * @return void
+     */
+    private function initParameterListCollector($attributes)
```
[the remark this answers is not in the document]
Comment (mringler) [reply src/Propel/Generator/Builder/Util/SchemaReader.php]: same as with return types, I assumed you don't want them in Propel, as they are usually missing. I actually like using them, so thank you for the update

**Review by stereomon — APPROVED** [review]
Looks good to me except that the tests could also make use of return-types but that's not a blocker for me.

When checks are green we can merge this.

**Comment by mringler** [comment]
> Looks good to me except that the tests could also make use of return-types but that's not a blocker for me.
> 
> When checks are green we can merge this.

Great! I have linted the tests and added return types, too.

**Pull request merged**